streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1301,20 +1301,19 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
         }
       }
 
-      currPull.get.flatten.catchAll {
-        case e @ Some(e1) =>
-          (finalizer.get.flatMap(_(Exit.fail(e1))) *> finalizer.set(_ => UIO.unit)).uninterruptible *> ZIO.fail(
-            e
-          )
-        case None =>
-          (finalizer.get.flatMap(_(Exit.succeed(()))) *>
-            finalizer.set(_ => UIO.unit)).uninterruptible *>
-            pullOuter *>
-            go(as, finalizer, currPull)
+      currPull.get.flatten.catchAllCause { c =>
+        Cause.sequenceCauseOption(c) match {
+          case Some(e) => Pull.halt(e)
```

Comment:
Hmm why did we eagerly execute the finalizer here?

Reply:
The only thing I didn't think completely through in this PR because I was too eager to put it out is the fact that we lose the error signal through the `Exit`. What do you think we should do?

Reply:
Ok, convinvced myself it wasn't needed 👍

Reply:
`finalizerRef` for the win!

Reply:
Good point regarding the error. Let's add another Ref[Option[Cause[E]]] that is set on error and used in the finalizer?

Reply:
No need, `finalizerRef` takes care of it.

Reply:
How can we craft a test that confirms this?

Reply:
> Good point regarding the error. Let's add another Ref[Option[Cause[E]]] that is set on error and used in the finalizer?

Also on which error? What if there are more than one?

Reply:
The `Exit[Any, Any]` required in the finalizer is the one that results from the `ZManaged#use` block. It's a bit hard to see it here, but this should result from the terminal operation consuming the stream.

So the Exit value here really depends if the error was allowed to fail the use block or not. Which is to say - if there were no recovery operators downstream, we should see the error propagate back to the `finalizerRef`.

You can test this by flatMapping elements in a stream to a stream that differentiates between success/failure in its finalizer.

Reply:
Exactly, this was my intuition too. I'll add the test.